keps/0019-package-api-versioning.md

```diff
@@ -2,64 +2,374 @@
+
+### Operator developer (developer)
+
+Someone that builds or maintains an _operator_.
+
+### Package
+
+An archive that contains all necessary files for running an _operator_.
+
+### Package version
+
+A single version identifier that represents all of the underlying exact versions or minimum version requirements for an _operator_:
+
+- KUDO API version
+- Application version
+- Operator revision
+- Kubernetes version
+
+### Package Registry
+
+Storage accessible through the network (most commonly through the internet) containg multiple packages.
+
+### Application
+
+An underlying technology managed by an _operator_, e.g. Apache Cassandra, Apache Kafka, etc.
+
+It is possible that an _operator_ is composed of multiple _applications_.
+
+### Application version (app version)
+
+The version of an _application_, e.g. 3.11.5 (for Apache Cassandra), 2.3.0 (for Apache Kafka), etc.
+
+Even though this version will more commonly follow [SemVer](https://semver.org/) for operators managing popular applications, it is ultimately outside of our control and therefore cannot be assumed to be [SemVer](https://semver.org/).
+
+Programs dealing with the application version can try to parse it as a structured specification like SemVer, but should be able to handle anything (i.e., fall-back to strings).
 
 ## Summary
 
-The different versions of components that describe an operator package are surfaced to users, so that they are able to install exactly specify which version can be installed. Furthermore, compatibility with existing packages for older APIs of KUDO is provided.
+The different versions of components that constitute a _package_ are surfaced to users and developers so that they are able to build, maintain, query, manage and publish operators.
 
 ## Motivation
 
-The dependencies between KUDO, operators packages and the application that is managed by the operator enforce the use of multiple versions to fully describe these dependencies. First, there is the dependency on a specific API version of KUDO. Also, a specific version of an application is provided by the package. And last, the operator itself may have multiple revisions. All of these versions need to be captured as part of the operator in a concise way that makes it easy for users to install specific operators for any of these versions.
+The dependencies between KUDO, _packages_ and _applications_ managed by _operators_ require the use of multiple versions to fully reify an operator.
+
+All of the different versions specified in the ["package version" section](#package-version) need to be captured as part of the operator in a concise way that makes it easy for operator users to install specific operators for any of the available versions, possibly based on dependency version constraints (can I install the Kafka Operator version X on my Kubernetes cluster version Y with KUDO version Z?), and simple for operator developers to build and maintain operators.
```

Comment:
'be captured as part of the operator' or 'be captured as part of the package'? or both? 🤔